### PR TITLE
Rework Combination class

### PR DESCRIPTION
**Pull request by kaminow**
The previous iteration of the `Combination` classes required the computation graph for each pose to be held in GPU memory, which will quickly overflow normal GPUs when using all-atom poses. The new version splits the gradient calculation such that the gradient for each pose is done separately and combined appropriately at the end, meaning that each computation graph can be freed from memory after use. The derivation for the math used in the different `Combination` subclasses can be found in the `README_COMBINATION.md` file.

General list of changes for each file:

**README_COMBINATION.md**
Math for separating out the gradients in the `Combination` classes

**mtenn/combination.py**
Each method for combining predictions has a `torch.autograd.Function`, which takes care of combining and assigning the gradients in the backward pass, and a `Combination` subclass that is essentially a wrapper around the `Function`

**mtenn/conversion_utils/e3nn.py**
* Update `import` statements
* Add `ComplexOnlyStrategy`

**mtenn/conversion_utils/schnet.py**
* Update `import` statements
* Add `ComplexOnlyStrategy`

**mtenn/model.py**
* Move all non-`Model` classes to their own files
* Update `GroupedModel` forward pass to work with new `Combination` setup
* `GroupedModel` now returns list of predictions for each pose in addition to the final prediction

**mtenn/readout.py**
Move all `Readout`-related code

**mtenn/representation.py**
Move all `Representation`-related code

**mtenn/Strategy.py**
* Move all `Strategy`-related code
* Add `ComplexOnlyStrategy` class that only predicts on the full input

### CODE REVIEW

File: mtenn/combination.py
Comment (hmacdope) [inline mtenn/combination.py:85]: `MaxCombination` has a `__repr__` and a `__str__`, either make one here or get rid of other one (s).

File: mtenn/combination.py
Comment (hmacdope) [inline mtenn/combination.py:289]: You could make the associated `_XXX_CombinationFunc` a class attribute of the `XXCombination class` like this:

```python
class BoltzmannCombination(Combination):
    combination_func = _BoltzmannCombinationFunc

...
    def forward(...):
        return self.combination_func.apply(
           *args, **kwargs, 
        )
```

and then you might be able to refactor `forward` into the base class

But you don't have to, this setup is more explicit.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small things, but looking really good. Addressing failing test(s) is main issue, but good that we are putting them in / making progress towards checking correctness.

File: mtenn/combination.py
Comment (kaminow) [reply mtenn/combination.py:85]: the reason I implemented it for that one is that `MaxCombination` has parameters that I wanted included in the `repr`. for the other two, the default `repr` that's implemented by `pytorch` is fine. I can still add it in though for them if you think it'd be better

File: mtenn/combination.py
Comment (kaminow) [reply mtenn/combination.py:289]: I think I'll leave it as is, just because the `_MaxCombinationFunc` has a slightly different signature than the others, so I'd have to do something slightly hacky there I think

**Comment by kaminow** [comment]
after playing around with the tests a bit, it seems like it's just a stochastic failure based on how the random data is initialized. two workarounds that I can think of are:
1. find a random seed that lets all the tests through as is, and trust that if the math gets messed up at some point then the tests will fail
2. adjust the parameters to the `np.allclose` call to be more lenient
@hmacdope do you have thoughts as to which would be better/preferable?

File: mtenn/combination.py
Comment (hmacdope) [reply mtenn/combination.py:87]: Nah thats fine as is.

File: mtenn/combination.py
Comment (hmacdope) [reply mtenn/combination.py:291]: yeah no worries, I'm happy with that.

**Review by hmacdope — APPROVED** [review]
Looks good @kaminow, merge when ready.

Lets hold off on a release until we sort out CI and env build issues.

**Comment by hmacdope** [comment]
@kaminow I have fixes a missing ase (?) dep in tests and updated some env files.

**Comment by hmacdope** [comment]
I will investigate, seems odd.

**Comment by kaminow** [comment]
@hmacdope after some investigation, it seems that there's some requirements broken for the 3.11 version of `pytorch_geometric`. it seems to be requiring cuda for some reason, while the builds for older Python versions don't, so for Python 3.11 an older version of `pytorch_geometric` is being installed, prior to when the `interaction_graph` was added to the model

**Comment by kaminow** [comment]
for posterity, this is the error I get when I try to run `mamba install pytorch_geometric=2.3.1`:
```
warning  libmamba Added empty dependency for problem type SOLVER_RULE_UPDATE
Could not solve for environment specs
The following package could not be installed
└─ pytorch_geometric 2.3.1**  is installable and it requires
   └─ pyg-lib 0.2.0  with the potential options
      ├─ pyg-lib 0.2.0 would require
      │  └─ triton with the potential options
      │     ├─ triton [1.1.2|2.0.0] would require
      │     │  └─ python >=3.10,<3.11.0a0 , which can be installed;
      │     ├─ triton 1.1.2 would require
      │     │  └─ python >=3.7,<3.8.0a0 , which can be installed;
      │     ├─ triton [1.1.2|2.0.0] would require
      │     │  └─ python >=3.8,<3.9.0a0 , which can be installed;
      │     ├─ triton [1.1.2|2.0.0] would require
      │     │  └─ python >=3.9,<3.10.0a0 , which can be installed;
      │     └─ triton 2.0.0 would require
      │        └─ pytorch * cuda* with the potential options
      │           ├─ pytorch [1.0.1|1.1.0|...|1.9.1] would require
      │           │  └─ python >=3.7,<3.8.0a0 , which can be installed;
      │           ├─ pytorch [1.10.0|1.10.1|...|1.9.1] would require
      │           │  └─ python >=3.8,<3.9.0a0 , which can be installed;
      │           ├─ pytorch [1.10.0|1.10.1|...|1.9.1] would require
      │           │  └─ python >=3.9,<3.10.0a0 , which can be installed;
      │           ├─ pytorch 1.11.0 would require
      │           │  └─ python >=3.10,<3.11.0a0 , which can be installed;
      │           ├─ pytorch [1.11.0|1.12.0|...|2.0.0] would require
      │           │  └─ __cuda, which is missing on the system;
      │           ├─ pytorch [1.0.1|1.1.0|...|1.9.1] would require
      │           │  └─ python >=3.6,<3.7.0a0 , which can be installed;
      │           ├─ pytorch [1.0.1|1.1.0|1.2.0|1.3.1] would require
      │           │  └─ python >=2.7,<2.8.0a0 , which can be installed;
      │           └─ pytorch 1.0.1 would require
      │              └─ cudatoolkit >=8.0,<8.1.0a0 , which does not exist (perhaps a missing channel);
      ├─ pyg-lib 0.2.0 would require
      │  └─ python >=3.10,<3.11.0a0 , which can be installed;
      ├─ pyg-lib 0.2.0 would require
      │  └─ python >=3.8,<3.9.0a0 , which can be installed;
      └─ pyg-lib 0.2.0 would require
         └─ python >=3.9,<3.10.0a0 , which can be installed.
```

**Comment by hmacdope** [comment]
@kaminow let me take a quick look on their feedstock.

**Comment by hmacdope** [comment]
> for posterity, this is the error I get when I try to run `mamba install pytorch_geometric=2.3.1` in a Python 3.11 env:
> 
> ```
> warning  libmamba Added empty dependency for problem type SOLVER_RULE_UPDATE
> Could not solve for environment specs
> The following package could not be installed
> └─ pytorch_geometric 2.3.1**  is installable and it requires
>    └─ pyg-lib 0.2.0  with the potential options
>       ├─ pyg-lib 0.2.0 would require
>       │  └─ triton with the potential options
>       │     ├─ triton [1.1.2|2.0.0] would require
>       │     │  └─ python >=3.10,<3.11.0a0 , which can be installed;
>       │     ├─ triton 1.1.2 would require
>       │     │  └─ python >=3.7,<3.8.0a0 , which can be installed;
>       │     ├─ triton [1.1.2|2.0.0] would require
>       │     │  └─ python >=3.8,<3.9.0a0 , which can be installed;
>       │     ├─ triton [1.1.2|2.0.0] would require
>       │     │  └─ python >=3.9,<3.10.0a0 , which can be installed;
>       │     └─ triton 2.0.0 would require
>       │        └─ pytorch * cuda* with the potential options
>       │           ├─ pytorch [1.0.1|1.1.0|...|1.9.1] would require
>       │           │  └─ python >=3.7,<3.8.0a0 , which can be installed;
>       │           ├─ pytorch [1.10.0|1.10.1|...|1.9.1] would require
>       │           │  └─ python >=3.8,<3.9.0a0 , which can be installed;
>       │           ├─ pytorch [1.10.0|1.10.1|...|1.9.1] would require
>       │           │  └─ python >=3.9,<3.10.0a0 , which can be installed;
>       │           ├─ pytorch 1.11.0 would require
>       │           │  └─ python >=3.10,<3.11.0a0 , which can be installed;
>       │           ├─ pytorch [1.11.0|1.12.0|...|2.0.0] would require
>       │           │  └─ __cuda, which is missing on the system;
>       │           ├─ pytorch [1.0.1|1.1.0|...|1.9.1] would require
>       │           │  └─ python >=3.6,<3.7.0a0 , which can be installed;
>       │           ├─ pytorch [1.0.1|1.1.0|1.2.0|1.3.1] would require
>       │           │  └─ python >=2.7,<2.8.0a0 , which can be installed;
>       │           └─ pytorch 1.0.1 would require
>       │              └─ cudatoolkit >=8.0,<8.1.0a0 , which does not exist (perhaps a missing channel);
>       ├─ pyg-lib 0.2.0 would require
>       │  └─ python >=3.10,<3.11.0a0 , which can be installed;
>       ├─ pyg-lib 0.2.0 would require
>       │  └─ python >=3.8,<3.9.0a0 , which can be installed;
>       └─ pyg-lib 0.2.0 would require
>          └─ python >=3.9,<3.10.0a0 , which can be installed.
> ```

Pinging @mikemhenry as well as I see he is a maintainer on the PYG feedstock

**Comment by hmacdope** [comment]
I am fairly sure this is due to the exact pin of `pyg-lib==0.2.0` in the `pyg` feedstock which is pulling down old pytorch versions. Tagging @hadim and @rusty1s? Perhaps they have some insight also. I will also confirm when on my linux box. Regardless, I think we are OK to push forward here and leave CI as indicating a failure.

**Review by hmacdope — APPROVED** [review]
LGTM

**Pull request merged**
